File: python/context_steward/text_optimizer.py
"""
Text Optimizer - Core optimization engine for Python.

Reduces token count through:
- Filler word removal
- Phrase simplification
- Redundancy elimination
- Smart truncation
"""

import re
from typing import Optional

try:
    import tiktoken
    HAS_TIKTOKEN = True
except ImportError:
    HAS_TIKTOKEN = False

from .types import OptimizeParams, OptimizationResult, OptimizationStrategy


# Default filler words to remove
FILLER_WORDS = [
    "very", "really", "quite", "rather", "somewhat", "fairly", "pretty",
    "just", "simply", "basically", "essentially", "generally", "typically",
    "obviously", "clearly", "certainly", "definitely", "absolutely",
    "completely", "totally", "entirely", "wholly", "actually", "literally"
]

# Phrase simplifications
PHRASE_REPLACEMENTS = [
    (r"in order to", "to"),
    (r"due to the fact that", "because"),
    (r"for the purpose of", "for"),
    (r"in the event that", "if"),
    (r"at this point in time", "now"),
    (r"in the near future", "soon"),
    (r"a large number of", "many"),
    (r"a small number of", "few"),
    (r"it should be noted that", ""),
    (r"it is important to note that", ""),
    (r"please be aware that", ""),
    (r"i would like to", "I'd like to"),
    (r"please kindly", "please"),
    (r"could you please", "please"),
    (r"would you be able to", "can you"),
    (r"i was wondering if", "can"),
    (r"as a matter of fact", ""),
    (r"at the present time", "now"),
    (r"despite the fact that", "although"),
    (r"in spite of the fact that", "although"),
    (r"has the ability to", "can"),
    (r"is able to", "can"),
    (r"make a decision", "decide"),
    (r"take into consideration", "consider"),
    (r"come to the conclusion", "conclude"),
    (r"give an explanation", "explain"),
]

# Aggressive-only patterns
AGGRESSIVE_REPLACEMENTS = [
    (r"\b(a|an|the)\s+(?=\w)", ""),
    (r",\s*which\s+", " that "),
]

# Transition words to remove in aggressive mode
TRANSITION_WORDS = [
    "however", "moreover", "furthermore", "additionally", "consequently",
    "therefore", "thus", "hence", "accordingly", "meanwhile", "nevertheless",
    "nonetheless", "alternatively", "specifically", "particularly"
]


class TextOptimizer:
    """Text optimization engine."""

    def __init__(
        self,
        preserve_terms: Optional[list[str]] = None,
        model: str = "gpt-4"
    ):
        self.preserve_terms = preserve_terms or []
        self.model = model
        self._encoder = None

    def _get_encoder(self):
        """Get or create tiktoken encoder."""
        if self._encoder is None and HAS_TIKTOKEN:
            try:
                self._encoder = tiktoken.encoding_for_model(self.model)
            except KeyError:
                self._encoder = tiktoken.get_encoding("cl100k_base")
        return self._encoder

    def count_tokens(self, text: str) -> int:
        """Count tokens in text."""
        encoder = self._get_encoder()
        if encoder:
            return len(encoder.encode(text))
        # Fallback: rough estimate
        return len(text) // 4

    def optimize(self, params: OptimizeParams) -> OptimizationResult:
        """Optimize text to reduce token count."""
        text = params.text
        strategy = params.strategy or OptimizationStrategy.BALANCED
        all_preserve_terms = list(set(self.preserve_terms + params.preserve_terms))
        
        original_tokens = self.count_tokens(text)
        techniques_applied = []

        # Apply strategy-based optimization
        if strategy == OptimizationStrategy.CONSERVATIVE:
            optimized = self._conservative_optimization(text, all_preserve_terms)
            techniques_applied = ["whitespace_cleanup", "minimal_filler_removal"]
        elif strategy == OptimizationStrategy.AGGRESSIVE:
            optimized = self._aggressive_optimization(text, all_preserve_terms)
            techniques_applied = ["filler_removal", "phrase_simplification", "article_removal", "transition_removal"]
        else:  # balanced
            optimized = self._balanced_optimization(text, all_preserve_terms)
            techniques_applied = ["filler_removal", "phrase_simplification"]

        # Truncate if max_tokens specified
        if params.max_tokens:
            current_tokens = self.count_tokens(optimized)
            if current_tokens > params.max_tokens:
                optimized = self._truncate_to_token_limit(optimized, params.max_tokens, all_preserve_terms)
                techniques_applied.append("truncation")

        # Preserve formatting if requested
        if params.preserve_formatting:
            optimized = self._preserve_basic_formatting(optimized, text)

        # Ensure preserved terms
        optimized = self._ensure_preserved_terms(optimized, text, all_preserve_terms)

        optimized_tokens = self.count_tokens(optimized)
        reduction_ratio = 1 - (optimized_tokens / original_tokens) if original_tokens > 0 else 0

        return OptimizationResult(
            original_text=text,
            optimized_text=optimized,
            original_tokens=original_tokens,
            optimized_tokens=optimized_tokens,
            reduction_ratio=reduction_ratio,
            strategy=strategy,
            preserved_terms=all_preserve_terms,
            techniques_applied=techniques_applied
        )

    def _conservative_optimization(self, text: str, preserve_terms: list[str]) -> str:
        """Minimal changes, safe for all content."""
        optimized = text
        
        # Remove extra whitespace only
        optimized = " ".join(optimized.split())
        
        # Remove only the most redundant words
        for word in ["very", "really", "quite", "rather"]:
            optimized = re.sub(rf"\b{word}\s+", "", optimized, flags=re.IGNORECASE)
        
        return optimized

    def _balanced_optimization(self, text: str, preserve_terms: list[str]) -> str:
        """Good reduction while maintaining readability."""
        optimized = text
        
        # Remove filler words
        for word in FILLER_WORDS:
            optimized = re.sub(rf"\b{word}\s+", "", optimized, flags=re.IGNORECASE)
        
        # Apply phrase simplifications
        for pattern, replacement in PHRASE_REPLACEMENTS:
            optimized = re.sub(pattern, replacement, optimized, flags=re.IGNORECASE)
        
        # Clean up extra spaces
        optimized = " ".join(optimized.split())
        
        return optimized

    def _aggressive_optimization(self, text: str, preserve_terms: list[str]) -> str:
        """Maximum reduction, may affect style."""
        optimized = self._balanced_optimization(text, preserve_terms)
        
        # Apply aggressive patterns
        for pattern, replacement in AGGRESSIVE_REPLACEMENTS:
            optimized = re.sub(pattern, replacement, optimized, flags=re.IGNORECASE)
        
        # Remove transition words
        for word in TRANSITION_WORDS:
            optimized = re.sub(rf"\b{word},?\s*", "", optimized, flags=re.IGNORECASE)
        
        # Abbreviate common phrases
        abbreviations = [
            (r"for example", "e.g."),
            (r"that is", "i.e."),
            (r"and so forth", "etc."),
            (r"with respect to", "re:"),
        ]
        for pattern, replacement in abbreviations:
            optimized = re.sub(pattern, replacement, optimized, flags=re.IGNORECASE)
        
        # Clean up
        optimized = " ".join(optimized.split())
        optimized = re.sub(r"\s*,\s*", ", ", optimized)
        optimized = re.sub(r"\s*\.\s*", ". ", optimized)
        
        return optimized.strip()

    def _truncate_to_token_limit(
        self, text: str, token_limit: int, preserve_terms: list[str]
    ) -> str:
        """Truncate text to fit token limit, prioritizing preserved terms."""
        sentences = re.split(r"(?<=[.!?])\s+", text)
        
        with_terms = []
        without_terms = []
        
        for sentence in sentences:
            has_term = any(
                term.lower() in sentence.lower() for term in preserve_terms
            )
            if has_term:
                with_terms.append(sentence)
            else:
                without_terms.append(sentence)
        
        result = ""
        current_tokens = 0
        
        for sentence in with_terms + without_terms:
            sentence_tokens = self.count_tokens(sentence)
            if current_tokens + sentence_tokens <= token_limit:
                result += (" " if result else "") + sentence
                current_tokens += sentence_tokens
            else:
                break
        
        return result or text[:token_limit * 4]

    def _preserve_basic_formatting(self, optimized: str, original: str) -> str:
        """Preserve paragraph breaks from original."""
        paragraphs = original.split("\n\n")
        if len(paragraphs) > 1:
            sentences = re.split(r"(?<=[.!?])\s+", optimized)
            sentences_per_para = max(1, len(sentences) // len(paragraphs))
            
            formatted = ""
            for i in range(0, len(sentences), sentences_per_para):
                para_sentences = sentences[i:i + sentences_per_para]
                formatted += " ".join(para_sentences) + "\n\n"
            
            return formatted.strip()
        
        return optimized

    def _ensure_preserved_terms(
        self, optimized: str, original: str, preserve_terms: list[str]
    ) -> str:
        """Ensure preserved terms weren't accidentally removed."""
        result = optimized
        
        for term in preserve_terms:
            pattern = re.compile(rf"\b{re.escape(term)}\b", re.IGNORECASE)
            if pattern.search(original) and not pattern.search(result):
                result += f" (Note: {term})"
        
        return result


# Patch: Add cleanup method to fix edge cases
def _cleanup_artifacts(text: str) -> str:
    """Clean up artifacts from optimization."""
    import re
    
    # Fix orphaned punctuation
    text = re.sub(r',\s*,', ',', text)
    text = re.sub(r'\s+,', ',', text)
    text = re.sub(r',\s+\.', '.', text)
    
    # Fix double spaces
    text = re.sub(r'\s+', ' ', text)
    
    # Fix "i. e." -> "i.e."
    text = re.sub(r'i\.\s*e\.', 'i.e.', text)
    text = re.sub(r'e\.\s*g\.', 'e.g.', text)
    
    # Fix sentence start after removal
    text = re.sub(r'\.\s+([a-z])', lambda m: '. ' + m.group(1).upper(), text)
    
    # Fix double verbs from pattern overlap
    text = re.sub(r'\bcan you could\b', 'could you', text, flags=re.IGNORECASE)
    text = re.sub(r'\bcan you please\b', 'please', text, flags=re.IGNORECASE)
    
    return text.strip()

# Monkey-patch into the class
TextOptimizer._cleanup_artifacts = staticmethod(_cleanup_artifacts)
_original_optimize = TextOptimizer.optimize

def _patched_optimize(self, params):
    result = _original_optimize(self, params)
    result.optimized_text = _cleanup_artifacts(result.optimized_text)
    result.optimized_tokens = self.count_tokens(result.optimized_text)
    result.reduction_ratio = 1 - (result.optimized_tokens / result.original_tokens) if result.original_tokens > 0 else 0
    return result

TextOptimizer.optimize = _patched_optimize

File: src/core/text-optimizer.ts
/**
 * Text Optimizer - Core optimization engine
 * 
 * Reduces token count through:
 * - Filler word removal
 * - Phrase simplification
 * - Redundancy elimination
 * - Smart truncation
 */

import type { 
  OptimizeParams, 
  OptimizationResult, 
  OptimizationStrategy,
  LLMAdapter 
} from '../types';

// Default filler words to remove
const FILLER_WORDS = [
  'very', 'really', 'quite', 'rather', 'somewhat', 'fairly', 'pretty',
  'just', 'simply', 'basically', 'essentially', 'generally', 'typically',
  'obviously', 'clearly', 'certainly', 'definitely', 'absolutely',
  'completely', 'totally', 'entirely', 'wholly', 'actually', 'literally'
];

// Phrase simplifications
const PHRASE_REPLACEMENTS: Array<{ pattern: RegExp; replacement: string; aggressive?: boolean }> = [
  { pattern: /in order to/gi, replacement: 'to' },
  { pattern: /due to the fact that/gi, replacement: 'because' },
  { pattern: /for the purpose of/gi, replacement: 'for' },
  { pattern: /in the event that/gi, replacement: 'if' },
  { pattern: /at this point in time/gi, replacement: 'now' },
  { pattern: /in the near future/gi, replacement: 'soon' },
  { pattern: /a large number of/gi, replacement: 'many' },
  { pattern: /a small number of/gi, replacement: 'few' },
  { pattern: /it should be noted that/gi, replacement: '' },
  { pattern: /it is important to note that/gi, replacement: '' },
  { pattern: /please be aware that/gi, replacement: '' },
  { pattern: /i would like to/gi, replacement: "I'd like to" },
  { pattern: /please kindly/gi, replacement: 'please' },
  { pattern: /could you please/gi, replacement: 'please' },
  { pattern: /would you be able to/gi, replacement: 'can you' },
  { pattern: /i was wondering if/gi, replacement: 'can' },
  { pattern: /as a matter of fact/gi, replacement: '' },
  { pattern: /at the present time/gi, replacement: 'now' },
  { pattern: /despite the fact that/gi, replacement: 'although' },
  { pattern: /in spite of the fact that/gi, replacement: 'although' },
  { pattern: /has the ability to/gi, replacement: 'can' },
  { pattern: /is able to/gi, replacement: 'can' },
  { pattern: /make a decision/gi, replacement: 'decide' },
  { pattern: /take into consideration/gi, replacement: 'consider' },
  { pattern: /come to the conclusion/gi, replacement: 'conclude' },
  { pattern: /give an explanation/gi, replacement: 'explain' },
  // Aggressive only
  { pattern: /\b(a|an|the)\s+(?=\w)/gi, replacement: '', aggressive: true },
  { pattern: /,\s*which\s+/gi, replacement: ' that ', aggressive: true },
];

// Transition words to remove in aggressive mode
const TRANSITION_WORDS = [
  'however', 'moreover', 'furthermore', 'additionally', 'consequently',
  'therefore', 'thus', 'hence', 'accordingly', 'meanwhile', 'nevertheless',
  'nonetheless', 'alternatively', 'specifically', 'particularly'
];

export class TextOptimizer {
  private adapter?: LLMAdapter;
  private preserveTerms: string[];
  private preservePatterns: RegExp[];

  constructor(config?: {
    adapter?: LLMAdapter;
    preserveTerms?: string[];
    preservePatterns?: RegExp[];
  }) {
    this.adapter = config?.adapter;
    this.preserveTerms = config?.preserveTerms || [];
    this.preservePatterns = config?.preservePatterns || [];
  }

  /**
   * Count tokens in text
   */
  countTokens(text: string): number {
    if (this.adapter) {
      return this.adapter.countTokens(text);
    }
    // Fallback: rough estimate (1 token ≈ 4 chars)
    return Math.ceil(text.length / 4);
  }

  /**
   * Optimize text to reduce token count
   */
  async optimize(params: OptimizeParams): Promise<OptimizationResult> {
    const {
      text,
      targetReduction = 0.3,
      preserveFormatting = false,
      preserveTerms = [],
      strategy = 'balanced',
      maxTokens
    } = params;

    const allPreserveTerms = [...this.preserveTerms, ...preserveTerms];
    const originalTokens = this.countTokens(text);
    
    let optimizedText = text;
    const techniquesApplied: string[] = [];

    // Apply strategy-based optimization
    switch (strategy) {
      case 'conservative':
        optimizedText = this.conservativeOptimization(optimizedText, allPreserveTerms);
        techniquesApplied.push('whitespace_cleanup', 'minimal_filler_removal');
        break;
      
      case 'aggressive':
        optimizedText = this.aggressiveOptimization(optimizedText, allPreserveTerms);
        techniquesApplied.push('filler_removal', 'phrase_simplification', 'article_removal', 'transition_removal');
        break;
      
      case 'balanced':
      default:
        optimizedText = this.balancedOptimization(optimizedText, allPreserveTerms);
        techniquesApplied.push('filler_removal', 'phrase_simplification');
        break;
    }

    // If max tokens specified, truncate if needed
    if (maxTokens) {
      const currentTokens = this.countTokens(optimizedText);
      if (currentTokens > maxTokens) {
        optimizedText = this.truncateToTokenLimit(optimizedText, maxTokens, allPreserveTerms);
        techniquesApplied.push('truncation');
      }
    }

    // Preserve formatting if requested
    if (preserveFormatting) {
      optimizedText = this.preserveBasicFormatting(optimizedText, text);
    }

    // Restore preserved terms if accidentally removed
    optimizedText = this.ensurePreservedTerms(optimizedText, text, allPreserveTerms);

    const optimizedTokens = this.countTokens(optimizedText);
    const reductionRatio = 1 - (optimizedTokens / originalTokens);

    return {
      originalText: text,
      optimizedText,
      originalTokens,
      optimizedTokens,
      reductionRatio,
      strategy,
      preservedTerms: allPreserveTerms,
      techniquesApplied
    };
  }

  /**
   * Conservative optimization - minimal changes
   */
  private conservativeOptimization(text: string, preserveTerms: string[]): string {
    let optimized = text;

    // Remove extra whitespace only
    optimized = optimized.replace(/\s+/g, ' ').trim();

    // Remove only the most redundant words
    const minimalFillers = ['very', 'really', 'quite', 'rather'];
    for (const word of minimalFillers) {
      const regex = new RegExp(`\\b${word}\\s+`, 'gi');
      optimized = optimized.replace(regex, '');
    }

    return optimized;
  }

  /**
   * Balanced optimization - good reduction while maintaining readability
   */
  private balancedOptimization(text: string, preserveTerms: string[]): string {
    let optimized = text;

    // Remove filler words
    for (const word of FILLER_WORDS) {
      const regex = new RegExp(`\\b${word}\\s+`, 'gi');
      optimized = optimized.replace(regex, '');
    }

    // Apply phrase simplifications (non-aggressive only)
    for (const { pattern, replacement, aggressive } of PHRASE_REPLACEMENTS) {
      if (!aggressive) {
        optimized = optimized.replace(pattern, replacement);
      }
    }

    // Clean up extra spaces
    optimized = optimized.replace(/\s+/g, ' ').trim();

    return optimized;
  }

  /**
   * Aggressive optimization - maximum reduction
   */
  private aggressiveOptimization(text: string, preserveTerms: string[]): string {
    let optimized = text;

    // Start with balanced optimization
    optimized = this.balancedOptimization(optimized, preserveTerms);

    // Apply aggressive phrase replacements
    for (const { pattern, replacement, aggressive } of PHRASE_REPLACEMENTS) {
      if (aggressive) {
        optimized = optimized.replace(pattern, replacement);
      }
    }

    // Remove transition words
    for (const word of TRANSITION_WORDS) {
      const regex = new RegExp(`\\b${word},?\\s*`, 'gi');
      optimized = optimized.replace(regex, '');
    }

    // Abbreviate common phrases
    const abbreviations = [
      { pattern: /for example/gi, replacement: 'e.g.' },
      { pattern: /that is/gi, replacement: 'i.e.' },
      { pattern: /and so forth/gi, replacement: 'etc.' },
      { pattern: /with respect to/gi, replacement: 're:' },
    ];

    for (const { pattern, replacement } of abbreviations) {
      optimized = optimized.replace(pattern, replacement);
    }

    // Clean up spaces and punctuation
    optimized = optimized.replace(/\s+/g, ' ');
    optimized = optimized.replace(/\s*,\s*/g, ', ');
    optimized = optimized.replace(/\s*\.\s*/g, '. ');
    optimized = optimized.trim();

    return optimized;
  }

  /**
   * Truncate text to fit token limit
   */
  private truncateToTokenLimit(text: string, tokenLimit: number, preserveTerms: string[]): string {
    const sentences = text.split(/(?<=[.!?])\s+/).filter(s => s.trim());
    
    // Separate sentences with and without preserved terms
    const withTerms: string[] = [];
    const withoutTerms: string[] = [];

    for (const sentence of sentences) {
      const hasPreservedTerm = preserveTerms.some(term =>
        sentence.toLowerCase().includes(term.toLowerCase())
      );
      
      if (hasPreservedTerm) {
        withTerms.push(sentence);
      } else {
        withoutTerms.push(sentence);
      }
    }

    // Build result prioritizing sentences with preserved terms
    let result = '';
    let currentTokens = 0;

    for (const sentence of [...withTerms, ...withoutTerms]) {
      const sentenceTokens = this.countTokens(sentence);
      if (currentTokens + sentenceTokens <= tokenLimit) {
        result += (result ? ' ' : '') + sentence;
        currentTokens += sentenceTokens;
      } else {
        break;
      }
    }

    return result || text.substring(0, tokenLimit * 4); // Fallback
  }

  /**
   * Preserve basic formatting from original
   */
  private preserveBasicFormatting(optimized: string, original: string): string {
    // Preserve paragraph breaks
    const originalParagraphs = original.split(/\n\s*\n/);
    if (originalParagraphs.length > 1) {
      const sentences = optimized.split(/(?<=[.!?])\s+/).filter(s => s.trim());
      const sentencesPerParagraph = Math.ceil(sentences.length / originalParagraphs.length);
      
      let formatted = '';
      for (let i = 0; i < sentences.length; i += sentencesPerParagraph) {
        const paragraphSentences = sentences.slice(i, i + sentencesPerParagraph);
        formatted += paragraphSentences.join(' ') + '\n\n';
      }
      return formatted.trim();
    }

    return optimized;
  }

  /**
   * Ensure preserved terms weren't accidentally removed
   */
  private ensurePreservedTerms(optimized: string, original: string, preserveTerms: string[]): string {
    let result = optimized;
    
    for (const term of preserveTerms) {
      const termRegex = new RegExp(`\\b${this.escapeRegex(term)}\\b`, 'gi');
      
      // Check if term exists in original but not in optimized
      if (termRegex.test(original) && !termRegex.test(result)) {
        // Find context in original and append
        const match = original.match(new RegExp(`[^.!?]*\\b${this.escapeRegex(term)}\\b[^.!?]*`, 'i'));
        if (match) {
          result += ` (Note: ${term})`;
        }
      }
    }
    
    return result;
  }

  /**
   * Escape string for use in regex
   */
  private escapeRegex(str: string): string {
    return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  }
}
